Testnet/Multisig/Multisig/MultisigContract.cs

```diff
@@ -0,0 +1,355 @@
+                        parameterDecoded = Serializer.ToUInt32(parameterEncoded);
+                        
+                        break;
+                    case 6: // Int32
+                        parameterDecoded = Serializer.ToInt32(parameterEncoded);
+                        
+                        break;
+                    case 7: // UInt64
+                        parameterDecoded = Serializer.ToUInt64(parameterEncoded);
+                        
+                        break;
+                    case 8: // Int64
+                        parameterDecoded = Serializer.ToInt64(parameterEncoded);
+                        
+                        break;
+                    case 9: // Address
+                        parameterDecoded = Serializer.ToAddress(parameterEncoded);
+                        
+                        break;
+                    case 10: // Byte[]
+                        parameterDecoded = parameterEncoded;
+                        
+                        break;
+                    case 11: // UInt128
+                        parameterDecoded = Serializer.ToUInt128(parameterEncoded);
+                        
+                        break;
+                    case 12: // Uint256
+                        parameterDecoded = Serializer.ToUInt256(parameterEncoded);
+                        
+                        break;
+                    default:
+                        Assert(false, "Unknown parameter type");
+                        break;
+                }
+                
+                int currentLength = parameters.Length;
+                Array.Resize(ref parameters, currentLength + 1);
+                parameters[currentLength] = parameterDecoded;
+            }
+        }
+        
+        ITransferResult result = this.Call(tx.Destination, tx.Value, tx.MethodName, parameters);
+
+        if (result?.Success ?? false)
+            Log(new Execution() { TransactionId = transactionId });
+        else
+            Log(new ExecutionFailure() { TransactionId = transactionId });
+
+        tx.Executed = true;
+
+        State.SetStruct($"{TransactionPrefix}:{transactionId}", tx);
+    }
+
+    public void ChangeRequirement(uint required)
+    {
+        EnsureWalletOnly();
+
+        Assert(required > 0, "Can't have a quorum of less than 1.");
+        Assert(required <= OwnerCount, "Can't set quorum higher than the number of owners.");
+
+        Required = required;
+
+        Log(new RequirementChange() { Requirement = required });
+    }
+
+    /// <summary>
+    /// In the nomenclature of the Gnosis MultisigWallet contract this is based on, the 'wallet' refers to the contract itself.
+    /// In other words, methods with this restriction are intended to only be called by the contract itself.
+    /// </summary>
+    private void EnsureWalletOnly()
+    {
+        Assert(Message.Sender == Message.ContractAddress, "Can only be called by the contract itself.");
```

Comment:
This is clever, so it looks like transactions will be submitted by an owner, then confirmed by every other owner, then finally will call this contract from itself to execute the public method. 

Do I have that right?

I'm sure you've got it covered but this might get tricky as members join, they'll need to loop through and confirm all pending transactions. 

This also leaves a bit of a loophole, if a member confirms, then is removed. `Required` will be deducted from but the confirmations of pending transactions will remain the same. So if `Required=10` PersonA confirms Transaction. It will have 1 confirmation, then they leave and now `Required=9` but that transaction's confirmations is still 1. Somebody new can join bringing `Required=10` back but there still is that 1 confirmation but a no longer approved owner.

Reply:
Yes, the transactions get confirmed by a configurable number of other owners. Once the final confirmation is reached the transaction is 'executed'. This currently supports arbitrary destination contracts and methods; it was tested with the enhanced standard token contract, for example.

The member removal is a bit of an issue, but the transactions are not intended to remain unconfirmed for very long under normal circumstances.